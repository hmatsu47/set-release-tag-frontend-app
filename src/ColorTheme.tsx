import { blue, purple } from "@suid/material/colors";
import { createTheme } from "@suid/material/styles";

export const ColorTheme = createTheme({
  palette: {
    primary: {
      main: localStorage.getItem("primaryColor")
        ? (localStorage.getItem("primaryColor") as string)
        : blue[700],
    },
    secondary: {
      main: localStorage.getItem("secondaryColor")
        ? (localStorage.getItem("secondaryColor") as string)
        : purple[500],
    },
  },
});
