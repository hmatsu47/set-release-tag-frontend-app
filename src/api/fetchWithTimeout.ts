export const fetchWithTimeout = async (
  method: string,
  url: string,
  object?: Object
) => {
  const controller = new AbortController();
  const timeout = setTimeout(() => {
    controller.abort();
  }, 10000);

  try {
    const request = object
      ? {
          // object の指定があれば body に JSON 形式でセットしてリクエスト
          method: method,
          headers: {
            Accept: "application/json",
            "Content-Type": "application/json;charset=utf8",
          },
          body: JSON.stringify(object),
          signal: controller.signal,
        }
      : {
          // なければ body なしでリクエスト
          method: method,
          signal: controller.signal,
        };
    const response = await fetch(url, request);
    // このアプリケーションでは Error を throw しない
    // if (!response.ok) {
    //   throw new Error(response.statusText);
    // }
    return response;
  } finally {
    clearTimeout(timeout);
  }
};
