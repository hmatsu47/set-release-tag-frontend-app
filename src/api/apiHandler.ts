import { fetchWithTimeout } from "./fetchWithTimeout";

// ベース URI
export const baseUri = "/api";

// API からデータ取得（GET）
export const getApiData = async (url: string) => {
  return await (await fetchWithTimeout("GET", url)).json();
};
// API からデータ取得（POST）
export const postApiData = async (url: string, object: Object) => {
  return await (await fetchWithTimeout("POST", url, object)).json();
};
