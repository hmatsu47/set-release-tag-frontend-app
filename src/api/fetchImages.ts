import { ErrorResponse, ImageItem } from "../type";
import { baseUri, getApiData } from "./apiHandler";
import {
  setImages,
  setMessage,
  setMessageSeverity,
  setSelectedTag,
} from "../signal";

export const fetchImages = async () => {
  const load = async (): Promise<void> => {
    const data: ImageItem[] | ErrorResponse = await getApiData(
      `${baseUri}/images`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setImages(undefined);
      setMessage((data as ErrorResponse).message);
      setMessageSeverity("error");
      return;
    }
    setImages(data as ImageItem[]);
    setSelectedTag(undefined);
    setMessage(undefined);
  };
  void load();
};
