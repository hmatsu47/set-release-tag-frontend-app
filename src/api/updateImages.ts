import { ErrorResponse, ImageItem, ImageTag } from "../type";
import { baseUri, postApiData } from "./apiHandler";
import {
  selectedTag,
  setImages,
  setMessage,
  setMessageSeverity,
  setSelectedTag,
} from "../signal";

export const updateImages = async () => {
  if (!selectedTag()) {
    return;
  }
  const load = async (): Promise<void> => {
    const data: ImageItem[] | ErrorResponse = await postApiData(
      `${baseUri}/images`,
      {
        tag: selectedTag(),
      } as ImageTag
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setImages(undefined);
      setMessage((data as ErrorResponse).message);
      setMessageSeverity("error");
      return;
    }
    setImages(data as ImageItem[]);
    setSelectedTag(undefined);
    setMessage(undefined);
  };
  void load();
};
