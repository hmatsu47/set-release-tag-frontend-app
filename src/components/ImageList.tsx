import { For, Show, onMount } from "solid-js";
import Box from "@suid/material/Box";
import Button from "@suid/material/Button";
import Paper from "@suid/material/Paper";
import Table from "@suid/material/Table";
import TableBody from "@suid/material/TableBody";
import TableCell from "@suid/material/TableCell";
import TableContainer from "@suid/material/TableContainer";
import TableHead from "@suid/material/TableHead";
import TableRow from "@suid/material/TableRow";
import Typography from "@suid/material/Typography";
import { fetchImages } from "../api/fetchImages";
import { images, setIsOpenedConfirm, setSelectedTag } from "../signal";
import { formatDateTimeDisplay } from "../formatDateTime";

export const ImageList = () => {
  onMount(async () => {
    if (images() == undefined) {
      await fetchImages();
    }
  });

  return (
    <Show when={images() !== undefined} fallback={<></>}>
      <Box
        sx={{
          width: "100%",
          minWidth: "1024px",
          display: "flex",
        }}
        aria-live="polite"
      >
        <Show
          when={images() !== null}
          fallback={
            <Typography variant="h6">
              リポジトリにイメージがありません
            </Typography>
          }
        >
          <TableContainer component={Paper}>
            <Table aria-label="simple table">
              <TableHead>
                <TableRow>
                  <TableCell>選択</TableCell>
                  <TableCell>イメージタグ</TableCell>
                  <TableCell>プッシュ日時</TableCell>
                  <TableCell>
                    サイズ
                    <br />
                    （MB）
                  </TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                <For each={images()} fallback={<></>}>
                  {(imageItem) => (
                    <TableRow>
                      <TableCell component="th" scope="row">
                        <Button
                          variant={
                            imageItem.tags.filter((n) => n === "release")
                              .length > 0
                              ? "contained"
                              : "outlined"
                          }
                          size="small"
                          color={
                            imageItem.tags.filter((n) => n === "release")
                              .length > 0
                              ? "primary"
                              : "inherit"
                          }
                          onClick={async (e) => {
                            if (
                              imageItem.tags.filter((n) => n === "release")
                                .length == 0
                            ) {
                              setSelectedTag(imageItem.tags[0]);
                              setIsOpenedConfirm(true);
                            }
                          }}
                          sx={{ textTransform: "none" }}
                          title={`${imageItem.tags.join(", ")} を選択`}
                        >
                          選択
                        </Button>
                      </TableCell>
                      <TableCell>{imageItem.tags.join(", ")}</TableCell>
                      <TableCell>
                        {formatDateTimeDisplay(new Date(imageItem.pushed_at))}
                      </TableCell>
                      <TableCell align="right">
                        {(
                          Math.round(
                            (imageItem.size / (1000 * 1000)) * Math.pow(10, 2)
                          ) / Math.pow(10, 2)
                        ).toLocaleString()}
                      </TableCell>
                    </TableRow>
                  )}
                </For>
              </TableBody>
            </Table>
          </TableContainer>
        </Show>
      </Box>
    </Show>
  );
};
