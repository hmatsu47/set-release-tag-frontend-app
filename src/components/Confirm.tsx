import { Show } from "solid-js";
import { useTheme } from "@suid/material";
import Backdrop from "@suid/material/Backdrop";
import Box from "@suid/material/Box";
import Button from "@suid/material/Button";
import Stack from "@suid/material/Stack";
import Typography from "@suid/material/Typography";
import { updateImages } from "../api/updateImages";
import { isOpenedConfirm, setIsOpenedConfirm } from "../signal";

export const Confirm = () => {
  const theme = useTheme();

  return (
    <Show when={isOpenedConfirm() === true} fallback={<></>}>
      <Backdrop
        sx={{ color: "#fff", zIndex: theme.zIndex.drawer + 1 }}
        open={isOpenedConfirm()}
      ></Backdrop>
      <Box
        sx={{
          position: "absolute",
          top: "25%",
          left: "50%",
          transform: "translate(-50%, -50%)",
          width: 400,
          bgcolor: theme.palette.background.paper,
          border: "2px solid #000",
          boxShadow: "24px",
          p: 4,
          zIndex: theme.zIndex.drawer + 1,
        }}
      >
        <Typography variant="subtitle1">
          本番リリース時に使用するイメージを選択します。
        </Typography>
        <Typography variant="subtitle1">よろしいですか？</Typography>
        <Box sx={{ paddingTop: "20px" }}></Box>
        <Stack spacing={4} direction="row" sx={{ justifyContent: "center" }}>
          <Button
            variant="contained"
            size="small"
            color="primary"
            onClick={async (e) => {
              setIsOpenedConfirm(false);
              await updateImages();
            }}
            title="はい"
          >
            はい
          </Button>
          <Button
            variant="outlined"
            size="small"
            color="inherit"
            onClick={() => {
              setIsOpenedConfirm(false);
            }}
            title="いいえ"
          >
            いいえ
          </Button>
        </Stack>
      </Box>
    </Show>
  );
};
