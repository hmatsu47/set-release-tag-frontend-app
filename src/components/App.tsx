import { Component } from "solid-js";
import Box from "@suid/material/Box";
import Stack from "@suid/material/Stack";
import { ThemeProvider } from "@suid/material";
import { ColorTheme } from "../ColorTheme";
import { ImageList } from "./ImageList";
import { Message } from "./Message";
import { TitleBar } from "./TitleBar";
import { Confirm } from "./Confirm";

export const App: Component = () => {
  return (
    <ThemeProvider theme={ColorTheme}>
      <TitleBar />
      <Box
        sx={{
          padding: "10px 10px 0 10px",
          width: "100%",
          minWidth: "1024px",
          display: "flex",
          // justifyContent: "center",
        }}
        aria-live="polite"
      >
        <Stack spacing={2} direction="column">
          <Message />
          <ImageList />
          <Confirm />
        </Stack>
      </Box>
    </ThemeProvider>
  );
};
