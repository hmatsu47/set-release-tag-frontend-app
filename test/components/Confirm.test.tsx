// @vitest-environment jsdom

import { describe, expect, test, vi } from "vitest";
import "vi-fetch/setup";
import { mockFetch, mockPost } from "vi-fetch";
import { fireEvent, render } from "@solidjs/testing-library";
import { formatSnapshot } from "../common/formatSnapshot";
import { Confirm } from "../../src/components/Confirm";
import { baseUri } from "../../src/api/apiHandler";
import {
  isOpenedConfirm,
  setIsOpenedConfirm,
  setSelectedTag,
} from "../../src/signal";
import { ImageTag } from "../../src/type";

describe("<Confirm />", () => {
  const confirmList = [
    {
      title: "images の API コールが行われる（「はい」ボタンクリック）",
      digest:
        "sha256:4c716d4cf211c7b7d2f3233c941771ad0507ea5bacf93b492766aa41ae9f720d",
      pushedAt: new Date(),
      repositoryName: "hoge",
      size: 10000,
      tags: ["latest", "release"],
      attachTag: "release",
      buttonYes: true,
      testFetch: true,
    },
    {
      title: "images の API コールが行われない（「いいえ」ボタンクリック）",
      digest:
        "sha256:4c716d4cf211c7b7d2f3233c941771ad0507ea5bacf93b492766aa41ae9f720d",
      pushedAt: new Date(),
      repositoryName: "fuga",
      size: 2000000,
      tags: ["hoge", "fuga", "release"],
      attachTag: "release",
      buttonYes: false,
      testFetch: false,
    },
  ];
  beforeEach(() => {
    mockFetch.clearAll();
  });
  afterEach(() => {
    vi.clearAllMocks();
  });
  confirmList.forEach((testCase) => {
    test(testCase.title, async () => {
      setSelectedTag(testCase.attachTag);
      setIsOpenedConfirm(true);
      const mock = mockPost(`${baseUri}/images`)
        .withHeaders([["Content-Type", "application/json;charset=utf8"]])
        .willResolve({ tag: testCase.attachTag } as ImageTag);
      const { findByTitle, unmount } = render(() => <Confirm />);
      // はいボタン
      const buttonYes = (await findByTitle("はい")) as HTMLInputElement;
      expect(buttonYes).toHaveTextContent("はい");
      // いいえボタン
      const buttonNo = (await findByTitle("いいえ")) as HTMLInputElement;
      expect(buttonNo).toHaveTextContent("いいえ");
      const modal = (await findByTitle("modal")) as HTMLElement;
      // css の名前が動的に変わるので固定値に置換
      const htmlModal = formatSnapshot(modal.innerHTML);
      expect(htmlModal).toMatchSnapshot();
      // ボタンクリック
      fireEvent.click(testCase.buttonYes ? buttonYes : buttonNo);
      // API 呼び出しチェック
      if (testCase.testFetch) {
        // とりあえず呼び出しが行われたことだけを確認
        expect(mock).toHaveFetched();
        expect(mock).toHaveFetchedWithBody(
          JSON.stringify({ tag: testCase.attachTag } as ImageTag)
        );
      }
      // モーダルクローズ
      expect(isOpenedConfirm()).toBe(false);
      unmount();
    });
  });
});
