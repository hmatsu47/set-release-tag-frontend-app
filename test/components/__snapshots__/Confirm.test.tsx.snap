// Vitest Snapshot v1

exports[`<Confirm /> > images の API コールが行われない（「いいえ」ボタンクリック） 1`] = `"<h6 class=\\"MuiTypography-root MuiTypography-subtitle1 MuiTypography-root css-xxxxxxxx\\">リリース時に使用するイメージを選択します。</h6><h6 class=\\"MuiTypography-root MuiTypography-subtitle1 MuiTypography-root css-xxxxxxxx\\">よろしいですか？</h6><div class=\\"MuiBox-root css-xxxxxxxx\\"></div><div class=\\"MuiStack-root css-xxxxxxxx\\"><button type=\\"button\\" title=\\"はい\\" tabindex=\\"0\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-root MuiButtonBase-root css-xxxxxxxx\\">はい<span class=\\"MuiTouchRipple-root css-xxxxxxxx\\"></span></button><button type=\\"button\\" title=\\"いいえ\\" tabindex=\\"0\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxxxx\\">いいえ<span class=\\"MuiTouchRipple-root css-xxxxxxxx\\"></span></button></div>"`;

exports[`<Confirm /> > images の API コールが行われる（「はい」ボタンクリック） 1`] = `"<h6 class=\\"MuiTypography-root MuiTypography-subtitle1 MuiTypography-root css-xxxxxxxx\\">リリース時に使用するイメージを選択します。</h6><h6 class=\\"MuiTypography-root MuiTypography-subtitle1 MuiTypography-root css-xxxxxxxx\\">よろしいですか？</h6><div class=\\"MuiBox-root css-xxxxxxxx\\"></div><div class=\\"MuiStack-root css-xxxxxxxx\\"><button type=\\"button\\" title=\\"はい\\" tabindex=\\"0\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-root MuiButtonBase-root css-xxxxxxxx\\">はい<span class=\\"MuiTouchRipple-root css-xxxxxxxx\\"></span></button><button type=\\"button\\" title=\\"いいえ\\" tabindex=\\"0\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxxxx\\">いいえ<span class=\\"MuiTouchRipple-root css-xxxxxxxx\\"></span></button></div>"`;
